shared/clock/clock.cpp

```diff
@@ -1,55 +1,63 @@
 
 #include "clock.hpp"
 
+#include <array>
 #include <chrono>
 
 namespace utility::clock {
 
-    struct clockStruct {
-        // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+    struct ClockData {
+        /// When the clock was last updated under the true time
         NUClear::base_clock::time_point last_update = NUClear::base_clock::now();
-        // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+        /// Our calculated time when the clock was last updated in simulated time
         NUClear::base_clock::time_point epoch = last_update;
-        // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
-        double rtf = 1.0;  // real time factor
+        /// The real time factor of the simulated clock
+        double rtf = 1.0;
     };
 
-    static clockStruct clockArray[2];
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+    static std::array<ClockData, 3> data;
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+    static std::atomic_int active = 0;
 
-    static std::atomic_int clockArrayIndex  = 0;
+    /// Convenience function to cast durations to the base clock duration type so the functions fit on one line
+    template <typename T>
+    NUClear::base_clock::duration dc(T&& t) {
+        return std::chrono::duration_cast<NUClear::base_clock::duration>(t);
+    }
 
-    void update_rtf(const double& new_rtf) {
-        auto now = NUClear::clock::now();  // Changed from base_clock
+    void update_rtf(const double& rtf) {
+        using namespace std::chrono;  // Using namespace is fine in a function scope
 
-        static std::atomic_int updateClockIndex = clockArrayIndex ^ 1;  // Switch to other clock in array
+        auto now = NUClear::base_clock::now();
 
-        // Since we are updating our rtf we need to advance our state so the new deltas will be calculated properly
-        // We set the variables in this specific order to mimise the error that will occur if the threading reads as we
-        // are writing By reducing the delta between state and now any changes in rtf will have a minimal change on
-        // delta. The old rtf is used since it is assume that since the last update the old rtf was in effect and
-        // only now with future calculations the new rtf will be used
+        // Set up a circular buffer index for the clock data
+        int c = active;
+        int n = (c + 1) % data.size();
```

Comment:
Is this short for `current` and `next`? Maybe
```suggestion
        int c = active;   // current clock index to read from
        int n = (c + 1) % data.size();   // next clock index to set
```

Reply:
Yeah current and next, just wanted the next line to be simple and easy to read